src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSourceTests.cs

```diff
@@ -0,0 +1,195 @@
+
+            var token = testSubject.Subscribe(mockSink.Object);
+
+            token.Should().NotBeNull();
+            token.Should().BeOfType<SinkManager>();
+        }
+
+        [TestMethod]
+        public void Refresh_NoSubscribedSinksNoError()
+        {
+            var testSubject = new SonarErrorListDataSource(mockTableManagerProvider.Object);
+
+            Action act = () => testSubject.RefreshErrorList();
+            act.Should().NotThrow();
+        }
+
+        [TestMethod]
+        public void Refresh_OnlySubscribedSinksAreNotified()
+        {
+            var mockSink1 = new Mock<ITableDataSink>();
+            var mockSink2 = new Mock<ITableDataSink>();
+            var testSubject = new SonarErrorListDataSource(mockTableManagerProvider.Object);
+
+            var sink1Token = testSubject.Subscribe(mockSink1.Object);
+            testSubject.Subscribe(mockSink2.Object);
+
+            // 1. Refresh -> both sinks notified
+            testSubject.RefreshErrorList();
+
+            CheckSinkWasNotified(mockSink1);
+            CheckSinkWasNotified(mockSink2);
+
+            // 2. Unregister one sink then refresh
+            mockSink1.Reset();
+            mockSink2.Reset();
+            sink1Token.Dispose();
+
+            testSubject.RefreshErrorList();
+
+            CheckSinkWasNotNotified(mockSink1);
+            CheckSinkWasNotified(mockSink2);
+        }
+
+        [TestMethod]
+        public void AddSinkManager_NewManagerIsNotifiedOfExistingFactories()
+        {
+            var testSubject = new SonarErrorListDataSource(mockTableManagerProvider.Object);
+
+            var factory1 = Mock.Of<ITableEntriesSnapshotFactory>();
+            var factory2 = Mock.Of<ITableEntriesSnapshotFactory>();
+            testSubject.AddFactory(factory1);
+            testSubject.AddFactory(factory2);
+
+            var mockSinkManager = new Mock<ISinkManager>();
+
+            testSubject.AddSinkManager(mockSinkManager.Object);
+
+            CheckFactoryWasAdded(mockSinkManager, factory1);
+            CheckFactoryWasAdded(mockSinkManager, factory2);
+        }
+
+        [TestMethod]
+        public void AddFactory_ExistingSinkManagersAreNotifiedOfNewFactory()
+        {
+            var testSubject = new SonarErrorListDataSource(mockTableManagerProvider.Object);
+
+            var mockManager1 = new Mock<ISinkManager>();
+            var mockManager2 = new Mock<ISinkManager>();
+            testSubject.AddSinkManager(mockManager1.Object);
+            testSubject.AddSinkManager(mockManager2.Object);
+
+            var mockFactory = new Mock<ITableEntriesSnapshotFactory>();
+
+            testSubject.AddFactory(mockFactory.Object);
+
+            CheckFactoryWasAdded(mockManager1, mockFactory.Object);
+            CheckFactoryWasAdded(mockManager2, mockFactory.Object);
+        }
+
+        [TestMethod]
+        public void RemoveSinkManager_ManagerIsNotLongerNotified()
```

Comment:
General design question about RemoveSinkManager, and also about RemoveFactory: I see that they are all called only as part of Dispose of IssueTagger.
So I assume that in the end, when VS decided that the document is closed, it disposes ITagger, which triggers the Remove chain, right?

In this case, maybe in the future design changes, it would be useful to just change the classes to implement IDisposable, and test only that Dispose is clearing everything correctly.

Reply:
Not quite. There is one `IssueTracker` per file/document, but each `IssueTracker` will have multiple `IssueTagger` instances (one per view). The `IssueTracker` doesn't call `RemoveIssueTracker` until it has no more `IssueTagger` instances (i.e. the last view on the document is closed).

`RemoveSinkManager` is not called when a tagger is disposed. It's called when the sink manager is disposed, is done by VS to tell us that the `ITableDataSink` is no longer valid/available (which in the case of the error list probably means VS is being shut down).

We should be able to simplify the code, though.